Add initial data model (#1677)
thin-edge/thin-edge.io: docs/src/architecture/data-model.md

# thin-edge Data Model

The **data model** identifies all data send or received from/to **thin-edge** and it's components, to interact with those.
For all data it defines format and explains behaviour.

## Telemetry Data

**Telemetry Data** are **measurements**, **events** and **alarms**. Each is defined by a set of data-elements and has it specific behaviour.

### Measurements
  * **measurements** carry values from physical **Sensors**[^1] or a device's **Domain Application**[^1];
    e.g. voltage and current of an electricity meter, or current state of the manufacturing control process
  * **elements of a measurement:**
    - `type name`, a string that identifies the measurement uniquely in context of a device
    - `samples`, carry a single value or set of values, all taken at a single point in time

>       - `values` can be a mix of numbers, strings or booleans

Author: We decided that the initial version of the data-model shall reflect what is currently implemented. So I removed all others data types than numbers (see https://github.com/thin-edge/thin-edge.io/pull/1677/commits/a83bae7b7a5985accbb0cdbf9210d47c50667266).

With later improvements of the data-model we plan to support also strings and boolean. However as you said, we will consider that there might be clouds/cloud-mappers that may not support all data types.

        - optionally each value can have a `type name`
      - `timestamp` (conform to ISO 8601), that indicates when values were sampled
    - TODO: not yet supported: list of units per sample's values

>     - TODO: not yet supported: source

Reviewer: If it is a "not yet implemented" then I would rather use the construction char, 🚧.

e.g.
🚧 (not yet implemented): source

  * **behaviour of a measurement:**
    - thin-edge does not store any historical sampled values for measurements
    - there is no initialization-value for measurements; i.E. a measurement is not visible on thin-edge before the 1st sample was sent to thin-edge

### Events
  * **events** are notifications that something happened on a device's environment or software system;
    e.g. a sensor[^1] detected something like a door has been closed,
    or a system notification that e.g. a user has started an ssh session
  * **elements of an event:**
    - `type name`, a string that identifies the event uniquely in context of a device
    - `event text`, that carries a human readable event-message
      - must be UTF-8 encoded

>     - `timestamp` (conform to ISO 8601), that indicates when the event has occured

Reviewer: Is it worth mentioning that the `timestamp` can be automatically added by thin-edge.io when creating the measurement? So for simple cases, the user does not need to worry about the ISO8601 timestamp format when creating events/measurements/alarms (if they don't need to)?

Author: Done. See https://github.com/thin-edge/thin-edge.io/pull/1677/commits/10b6e223ac06eaf8d7dd3a47edd11df711bbc477

    - an event can optionally contain any custom-specific extra-information
    - TODO: not yet supported: attachements
  * **behaviour of an event:**
    - thin-edge does not store any historical occurrences for events

### Alarms
  * **alarms** are notifications about some critical behaviour of the device's environment or software system;
    e.g. when a temperature sensor detects a temperature went out of its valid range
  * **elements of an alarm:**
    - `type name`, a string that identifies the alarm uniquely in context of a device
    - alarm text, that carries an alarm message
      - must be UTF-8 encoded
    - severity, that is one of  `critical`, `major`, `minor` or `warning`
    - `timestamp` (conform to ISO 8601), that indicates when the alarm has occured

>     - an alarm can optionally contain any custom-specific extra-information

Reviewer: ```suggestion
    - an alarm can optionally contain additional custom information
```

  * **behaviour of an alarm:**
    - thin-edge does not store any historical occurrences for alarms

>     - **alarms** are stateful; i.E. once raised, an **alarm** is active until it was explicitely cleared by the device's software or the cloud

Reviewer: ```suggestion
    - **alarms** are stateful; i.e. once raised, an **alarm** is active until it was explicitly cleared by the device's software or the cloud
```

[^1]: details see "Domain Model" appendix [Device Domain](./domain-model.md#device-overview) -->

## Use of MQTT

> **thin-edge** expects the MQTT broker [mosquitto](https://mosquitto.org/) being installed on the device.

Reviewer: It will be better to keep it a bit more generic as we will be supported external mqtt brokers in the future, so the broker will not be installed on the device.

```suggestion
**thin-edge** expects the MQTT broker [mosquitto](https://mosquitto.org/) to be available on the device.
```

**thin-edge** uses **mosquitto** to consume and provide telemetry data. All telemetry data are reflected with specific MQTT topics and payload in JSON format.

> **thin-edge** assumes **mosquitto** is configured in a secure manner, to avoid any inappropriate access to **thin-edge** topics and payload.

Reviewer: It might be worth providing a link to the mosquitto ACL docs?

https://mosquitto.org/man/mosquitto-conf-5.html

Or even the optional mosquitto plugin. https://mosquitto.org/documentation/dynamic-security/

Author: Done. See https://github.com/thin-edge/thin-edge.io/pull/1677/commits/bfa02a8feda5bebdd6a2800534d0b41b34b2931d

Any malicious access to the broker can hazard **thin-edge** and all connected devices.

### Telemetry Data

> All telemetry data (**Measurments**, **Events**, **Alarms**) are reflected with MQTT topics, where each has it's specific subtopic (e.g. `tedge/measurements` or `tedge/events`).

Reviewer: We should also add a reference to the topic structure for child devices. Not sure if here is the best place, but somewhere on this page.

Author: Done. See https://github.com/thin-edge/thin-edge.io/pull/1677/commits/6a3b1408427ca59ba4750f2eb54070b57addd79c

# thin-edge Data Model

The **data model** identifies all data send or received from/to **thin-edge** and its components, to interact with those.
For all data it defines format and explains behaviour.

## Telemetry Data

**Telemetry Data** consists of **measurements**, **events** and **alarms**. Each is defined by a set of data-elements, each with specific behaviour.

### Measurements
**measurements** carry values from physical **Sensors**[^1] or a device's **Domain Application**[^1];
e.g. voltage and current of an electricity meter, or current state of the manufacturing control process

A measurement can carry a **single value**, or **muliple values** all taken at a single point in time.
```javascript
   {
     // example for a single-value messurement

>      "temperature":               // 'type_name' of that measurement

Reviewer: We should rename `type_name` as we will be supporting the `type` field in the future (currently it is only added when sending measurements to Cumulocity).

Author: Now that "type_name" is changed to "name".

                    25.3,         // 'value' of that measurement
     "time": "2020-10-15T05:30:47+00:00",  // optional 'timestamp' of that measurement
   }
```
```javascript
   {
     // example for a multi-value measurement
     "current": {                // 'type_name' of that measurement
        "L1": 9.5,               // the 1st 'value' of that measurement, named as "L1"
        "L2": 1.3                // the 2nd 'value' of that measurement, named as "L2"
        // ...even more values can occur
    },
     "time": "2020-10-15T05:30:47+00:00",  // optional 'timestamp' of that measurement
   }
```

|Reference  |Description|
| --------- | --------- |
|`type_name`  |a string that identifies the measurement uniquely in context of the device|
|`value`      |the value that was sampled; can be named (especially in context of a multi-value measurement) or unnamed; must be an integer or floating point number|
|`timestamp`  |optional time that indicates when values were sampled; when not provided, thin-edge.io uses the current system time as the time of the sample; when provided must be conform to ISO 8601|

>   * **behaviour of a measurement:**
>     - thin-edge does not store any historical sampled values for measurements
>     - there is no initialization value for measurements; i.e. a measurement is not visible on thin-edge before the 1st sample was sent to thin-edge

Reviewer: ```suggestion
**behaviour of a measurement**
- thin-edge does not store any historical sampled values for measurements
- there is no initialization value for measurements; i.e. a measurement is not visible on thin-edge before the 1st sample was sent to thin-edge
```

Reviewer: The behaviour section header does not need to be a list item itself (only its contents)

TODO: add somehow "an alarm can optionally contain additional custom information"

  * **behaviour of an alarm:**
    - thin-edge does not store any historical occurrences for alarms
    - **alarms** are stateful; i.e. once raised, an **alarm** is active until it was explicitly cleared by the device's software or the cloud

[^1]: details see "Domain Model" appendix [Device Domain](./domain-model.md#device-overview) -->

## Use of MQTT

**thin-edge** expects the MQTT broker [mosquitto](https://mosquitto.org/) to be available on the device.
**thin-edge** uses **mosquitto** to consume and provide telemetry data. All telemetry data are reflected with specific MQTT topics and payload in JSON format.

**thin-edge** assumes **mosquitto** is configured in a secure manner, to avoid any inappropriate access to **thin-edge** topics and payload.
Any malicious access to the broker can hazard **thin-edge** and all connected devices. Mosquitto provides a wide range of authentication and access control options. For more details see _Authentication_ and _ACL_ (Access Control List) in [mosquitto documentation](https://mosquitto.org/man/mosquitto-conf-5.html).

> ### Telemetry Data

Reviewer: I'm not sure if this section is really required now that the example of each element is shown in the section at the top of the page.

What if we move the missing details (e.g. topic name, retain etc.) from this section to the top? Or does that couple the model too much to MQTT?

Author: Sections "Telemetry Data" and "Use of MQTT" are reordered and partially merged now:
- put section 'Use of MQTT' before section 'Telemetry Data'
- merge telemetry specifics from section 'MQTT' into 'Telemetry Data'

Details see https://github.com/thin-edge/thin-edge.io/pull/1677/commits/0795970fa016ac62b302d18066a351c715037e70

> All telemetry data (**Measurements**, **Events**, **Alarms**) are reflected with MQTT topics, where each has it's specific subtopic (e.g. `tedge/measurements` or `tedge/events`).

Reviewer: ```suggestion
All telemetry data (**Measurements**, **Events**, **Alarms**) are reflected with MQTT topics, where each has its specific subtopic (e.g. `tedge/measurements`, `tedge/events`, `tedge/alarms` etc.).
```
